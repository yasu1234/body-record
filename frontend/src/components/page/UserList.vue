<script setup>
import { ref, onMounted } from "vue";
import { useRouter, onBeforeRouteUpdate } from "vue-router";
import axios from "axios";
import Cookies from "js-cookie";

import TabMenu from "../layout/TabMenu.vue";
import Header from "../layout/Header.vue";
import UserCard from "../layout/UserCard.vue";
import ListPage from "../layout/ListPage.vue";
import SearchButton from "../atom/SearchButton.vue";

const router = useRouter();

const keyword = ref('');
const isDisplayOnlySupport = ref(false);
const isLogin = ref(false);
const searchResult = ref([]);

const pageCount = ref(1);
const pageNum = ref(1);

onMounted(() => {
  searchUser();
});

onBeforeRouteUpdate(async (to, from) => {
  keyword.value = to.query.keyword;
  searchUser();
});

const targetSearch = () => {
  const query = {};

  if (keyword.value.trim() !== "") {
    query.keyword = keyword.value;
  }

  router.push({ path: "/users", query: query });
};

const searchUser = async () => {
  try {
    const res = await axios.get(
      import.meta.env.VITE_APP_API_BASE + "/api/v1/users",
      {
        headers: {
          "access-token": Cookies.get("accessToken"),
          client: Cookies.get("client"),
          uid: Cookies.get("uid"),
        },
        params: {
          keyword: keyword.value,
          page: pageNum,
        },
      }
    );

    if (res.data && res.data.total_page) {
      pageCount.value = res.data.total_page;
    } else {
      pageCount.value = 1;
    }

    searchResult.value = [];

    for (let item of res.data.users) {
      searchResult.value.push(item);
    }
  } catch (error) {
    searchResult.value = [];
  }
};

const updatePaginateItems = function (page) {
  pageNum.value = page;
  searchUser();
};

const userSelect = (item) => {
  router.push({ name: "OtherRecordList", params: { id: item.id } });
};
</script>

<template>
  <Header />
  <TabMenu :currentId="2" />
  <div class="user-search-container">
    <input
      type="text"
      id="keyword"
      name="keywordName"
      placeholder="名前で検索"
      v-model="keyword"
    />
    <div class="search-check">
      <input type="checkbox" id="statusSelect" v-model="isDisplayOnlySupport" />
      <label>応援しているユーザーのみ表示</label>
    </div>
    <div class="search-button-area">
      <SearchButton @searchButtonClick="targetSearch" />
    </div>
  </div>
  <div class="user-search-result">
    <div
      v-for="user in searchResult"
      class="user-card"
      @click="userSelect(user)"
    >
      <UserCard :user="user" />
    </div>
  </div>
  <div class="user-list-page">
    <ListPage
      :pageCount="pageCount"
      v-model="pageNum"
      @changePage="updatePaginateItems"
    />
  </div>
</template>

<style scoped>
.user-search-container {
  width: 700px;
  margin: 0 auto;
  padding: 20px;
  background-color: #ffffff;
  border: 1.5px solid #46c443;
  border-radius: 5px;
  margin-top: 20px;
}

input[type="text"] {
  width: 100%;
  padding: 12px 12px;
  margin: 8px 0;
  box-sizing: border-box;
  border: 1px solid #ccc;
  border-radius: 4px;
}

input[type="checkbox"] {
  border-radius: 0;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}
input[type="checkbox"] {
  position: relative;
  width: 16px;
  height: 16px;
  border: 1px solid #000;
  vertical-align: -5px;
}

input[type="checkbox"]:checked:before {
  position: absolute;
  top: 1px;
  left: 4px;
  transform: rotate(50deg);
  width: 4px;
  height: 8px;
  border-right: 2px solid #000;
  border-bottom: 2px solid #000;
  content: "";
}
.search-check {
  margin-top: 20px;
}
.user-search-result {
  margin-top: 20px;
}
.search-button-area {
  text-align: center;
  margin-top: 20px;
}
.user-card {
  width: 100%;
  max-width: 600px;
  margin: 0 auto;
  margin-bottom: 20px;
  margin-top: 10px;
  position: relative;
  box-shadow: 0px 1px 3px rgba(0, 0, 0, 0.18);
  border: 1px solid #ccc;
  border-radius: 5px;
  background-color: #ffffff;
  cursor: pointer;
}
.user-list-page {
  margin-top: 50px;
}
@media screen and (max-width: 768px) {
  .user-search-container {
    width: auto;
    margin-left: 20px;
    margin-right: 20px;
    padding: 20px;
    background-color: #ffffff;
    border: 1px solid #46c443;
    border-radius: 5px;
    margin-top: 20px;
  }
}
</style>
