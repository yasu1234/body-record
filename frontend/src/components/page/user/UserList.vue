<script setup>
import { ref, onMounted } from "vue";
import { useRouter, useRoute, onBeforeRouteUpdate } from "vue-router";
import { axiosInstance, setupInterceptors } from "../../../js/axios.js";

import TabMenu from "../../layout/TabMenu.vue";
import Header from "../../layout/Header.vue";
import UserCard from "../../layout/UserCard.vue";
import ListPage from "../../layout/ListPage.vue";
import SearchButton from "../../atom/SearchButton.vue";
import ResultEmpty from "../../atom/ResultEmpty.vue";

const route = useRoute();
const router = useRouter();
setupInterceptors(router);

const keyword = ref("");
const isDisplayOnlySupport = ref(false);
const isLogin = ref(false);
const searchResult = ref([]);
const pageCount = ref(1);
const page = ref(1);

onMounted(() => {
  setQuery(route.query.keyword, route.query.onlySupport, route.query.page);
  searchUser();
});

onBeforeRouteUpdate(async (to, from) => {
  setQuery(to.query.keyword, to.query.onlySupport, to.query.page);
  searchUser();
});

const targetSearch = () => {
  const query = {};

  if (keyword.value.trim() !== "") {
    query.keyword = keyword.value;
  }

  if (isDisplayOnlySupport.value) {
    query.onlySupport = true;
  }

  router.push({ path: "/users", query: query });
};

const setQuery = (keywordParam, onlySupportParam, pageParam) => {
  if (keywordParam != null) {
    keyword.value = keywordParam;
  } else {
    keyword.value = "";
  }
  if (onlySupportParam != null && onlySupportParam) {
    isDisplayOnlySupport.value = true;
  } else {
    isDisplayOnlySupport.value = false;
  }
  if (pageParam != null) {
    page.value = pageParam;
  } else {
    page.value = 1;
  }
};

const searchUser = async () => {
  try {
    const res = await axiosInstance.get("/api/v1/users", {
      params: {
        keyword: keyword.value,
        page: page.value,
        isSupportOnly: isDisplayOnlySupport.value,
      },
    });

    if (res.data && res.data.total_page) {
      pageCount.value = res.data.total_page;
    } else {
      pageCount.value = 1;
    }

    searchResult.value = [];

    for (let item of res.data.users) {
      searchResult.value.push(item);
    }
  } catch (error) {
    searchResult.value = [];
  }
};

const updatePaginateItems = function (page) {
  pageNum.value = page;
  searchUser();
};

const userSelect = (item) => {
  router.push({ name: "OtherRecordList", params: { id: item.user.id } });
};
</script>

<template>
  <Header />
  <TabMenu :currentId="2" />
  <div class="user-search-container">
    <input
      type="text"
      id="keyword"
      name="keywordName"
      placeholder="名前で検索"
      v-model="keyword"
    />
    <div class="mt-5">
      <input
        type="checkbox"
        id="statusSelect"
        v-model="isDisplayOnlySupport"
        class="w-4 h-4 text-blue-600 bg-gray-100 border-gray-300 rounded focus:ring-blue-500"
      />
      <label class="ml-2.5">応援しているユーザーのみ表示</label>
    </div>
    <div class="search-button-area">
      <SearchButton @searchButtonClick="targetSearch" />
    </div>
  </div>
  <div class="mt-5 pb-5">
    <div v-if="searchResult.length > 0">
      <div
        v-for="user in searchResult"
        class="user-card"
        @click="userSelect(user)"
      >
        <UserCard :user="user" />
      </div>
      <div class="mt-12">
        <ListPage
          :pageCount="pageCount"
          v-model="page"
          @changePage="updatePaginateItems"
        />
      </div>
    </div>
    <div v-else>
      <ResultEmpty class="mx-5" />
    </div>
  </div>
</template>

<style scoped>
.user-search-container {
  width: 700px;
  margin: 0 auto;
  padding: 20px;
  background-color: #ffffff;
  border: 1.5px solid #46c443;
  border-radius: 5px;
  margin-top: 20px;
}
input[type="text"] {
  width: 100%;
  padding: 12px 12px;
  margin: 8px 0;
  box-sizing: border-box;
  border: 1px solid #ccc;
  border-radius: 4px;
}
.search-button-area {
  text-align: center;
  margin-top: 20px;
}
.user-card {
  width: 100%;
  max-width: 600px;
  margin: 0 auto;
  margin-bottom: 20px;
  margin-top: 10px;
  position: relative;
  box-shadow: 0px 1px 3px rgba(0, 0, 0, 0.18);
  border: 1px solid #ccc;
  border-radius: 5px;
  background-color: #ffffff;
  cursor: pointer;
}

@media screen and (max-width: 768px) {
  .user-search-container {
    width: auto;
    margin-left: 20px;
    margin-right: 20px;
    padding: 20px;
    background-color: #ffffff;
    border: 1px solid #46c443;
    border-radius: 5px;
    margin-top: 20px;
  }
  .user-card {
    width: auto;
    margin-left: 20px;
    margin-right: 20px;
  }
}
</style>
