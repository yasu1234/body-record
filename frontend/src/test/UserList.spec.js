import { shallowMount } from "@vue/test-utils";
import UserList from "@/components/page/user/UserList.vue";

jest.mock("axios");
jest.mock("vue-router", () => {
  const router = {
    push: jest.fn(),
    pathname: "/",
    route: "",
    asPath: "",

    events: {
      on: jest.fn(),
      off: jest.fn(),
      emit: jest.fn(),
    },
    onBeforeRouteUpdate: jest.fn(),
  };
  return {
    useRouter: jest.fn().mockReturnValue(router),
    useRoute: jest.fn().mockReturnValue({
      params: {
        id: 1,
      },
      query: {
        keyword: "",
        onlySupport: false,
        page:1
      }
    }),
    onBeforeRouteUpdate: jest.fn()
  };
});

jest.mock("primevue/usetoast", () => ({
  useToast: jest.fn(() => ({
    add: jest.fn(),
    removeAll: jest.fn(),
  })),
}));

jest.mock("@/js/axios.js", () => ({
  axiosInstance: {
    get: jest.fn(),
    interceptors: {
      request: { use: jest.fn() },
      response: { use: jest.fn() },
    },
  },
  setupInterceptors: jest.fn(),
}));

describe("UserList", () => {
  let wrapper;
  let mockedAxiosInstance;

  beforeEach(() => {
    jest.mock("@/js/axios.js", () => ({
      axiosInstance: {
        get: jest.fn(),
        interceptors: {
          request: { use: jest.fn() },
          response: { use: jest.fn() },
        },
      },
      setupInterceptors: jest.fn(),
    }));
  
    // モックされたaxiosInstanceを取得
    const axiosModule = require("@/js/axios.js");
    mockedAxiosInstance = axiosModule.axiosInstance;
  
    // onMounted内でaxios.getが呼ばれるため、このタイミングでモック設定
    mockedAxiosInstance.get.mockResolvedValue({
      data: { users: [{id: 1, name: "AAA"}, {id:2, name: "BBB"}] , total_page: 2}
    });

    wrapper = shallowMount(UserList, {
      global: {
        stubs: ["router-link", "router-view"],
      },
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("SnapShotテスト初期表示(データあり)", () => {
    expect(wrapper.element).toMatchSnapshot();
  });
});

describe("UserList", () => {
    let wrapper;
    let mockedAxiosInstance;
  
    beforeEach(() => {
      jest.mock("@/js/axios.js", () => ({
        axiosInstance: {
          get: jest.fn(),
          interceptors: {
            request: { use: jest.fn() },
            response: { use: jest.fn() },
          },
        },
        setupInterceptors: jest.fn(),
      }));
    
      // モックされたaxiosInstanceを取得
      const axiosModule = require("@/js/axios.js");
      mockedAxiosInstance = axiosModule.axiosInstance;
    
      // onMounted内でaxios.getが呼ばれるため、このタイミングでモック設定
      mockedAxiosInstance.get.mockResolvedValue({
        data: { }
      });
  
      wrapper = shallowMount(UserList, {
        global: {
          stubs: ["router-link", "router-view"],
        },
      });
    });
  
    afterEach(() => {
      jest.clearAllMocks();
    });
  
    it("SnapShotテスト初期表示(データなし)", () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });
