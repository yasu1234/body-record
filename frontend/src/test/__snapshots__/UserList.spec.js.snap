// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserList SnapShotテスト初期表示(データあり) 1`] = `
<div
  data-v-app=""
>
  
  <header-stub />
  <tab-menu-stub
    currentid="3"
  />
  <div
    class="user-search-container"
  >
    <p
      class="font-bold"
    >
      ユーザーの絞り込み
    </p>
    <div
      class="mt-2.5"
    >
      <input
        name="keywordName"
        placeholder="名前で検索"
        type="text"
      />
    </div>
    <div
      class="mt-5"
    >
      <input
        class="w-4 h-4 text-blue-600 bg-gray-100 border-gray-300 rounded focus:ring-blue-500"
        type="checkbox"
      />
      <label
        class="ml-2.5"
      >
        応援しているユーザーのみ表示
      </label>
    </div>
    <div
      class="search-button-area"
    >
      <search-button-stub />
    </div>
  </div>
  <div
    class="py-8"
  >
    <div>
      <div
        class="select-box"
      >
        
        <div
          class="text-sm rounded-lg block p-3"
        >
          <radio-button-stub
            inputid="1"
            modelvalue="直近で作成された順"
            pt="[object Object]"
            value="直近で作成された順"
          />
          <label
            class="ml-2"
          >
            直近で作成された順
          </label>
        </div>
        <div
          class="text-sm rounded-lg block p-3"
        >
          <radio-button-stub
            inputid="2"
            modelvalue="直近で作成された順"
            pt="[object Object]"
            value="応援しているユーザーが多い順"
          />
          <label
            class="ml-2"
          >
            応援しているユーザーが多い順
          </label>
        </div>
        
      </div>
      <p
        class="text-center font-bold mt-8"
      >
        合計0件
      </p>
      
      <div
        class="user-card"
      >
        <user-card-stub
          user="[object Object]"
        />
      </div>
      <div
        class="user-card"
      >
        <user-card-stub
          user="[object Object]"
        />
      </div>
      
      <div
        class="mt-12"
      >
        <list-page-stub
          modelvalue="1"
          pagecount="2"
        />
      </div>
    </div>
    <!--v-if-->
    <!--v-if-->
  </div>
  
</div>
`;

exports[`UserList SnapShotテスト初期表示(データなし) 1`] = `
<div
  data-v-app=""
>
  
  <header-stub />
  <tab-menu-stub
    currentid="3"
  />
  <div
    class="user-search-container"
  >
    <p
      class="font-bold"
    >
      ユーザーの絞り込み
    </p>
    <div
      class="mt-2.5"
    >
      <input
        name="keywordName"
        placeholder="名前で検索"
        type="text"
      />
    </div>
    <div
      class="mt-5"
    >
      <input
        class="w-4 h-4 text-blue-600 bg-gray-100 border-gray-300 rounded focus:ring-blue-500"
        type="checkbox"
      />
      <label
        class="ml-2.5"
      >
        応援しているユーザーのみ表示
      </label>
    </div>
    <div
      class="search-button-area"
    >
      <search-button-stub />
    </div>
  </div>
  <div
    class="py-8"
  >
    <!--v-if-->
    <div>
      <result-empty-stub
        class="mx-5"
      />
    </div>
    <!--v-if-->
  </div>
  
</div>
`;
